protocols/rendezvous/src/behaviour.rs

```diff
@@ -0,0 +1,846 @@
+                        rendezvous_node: peer_id,
+                        namespace,
+                        error,
+                    }),
+                )]
+            }
+            OutEvent::UnregisterRequested(namespace) => {
+                self.registrations.remove(namespace.clone(), peer_id);
+
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::PeerUnregistered {
+                        peer: peer_id,
+                        namespace,
+                    },
+                )]
+            }
+            OutEvent::DiscoverRequested {
+                namespace,
+                cookie,
+                limit,
+            } => match self.registrations.get(namespace, cookie, limit) {
+                Ok((registrations, cookie)) => {
+                    let discovered = registrations.cloned().collect::<Vec<_>>();
+
+                    vec![
+                        NetworkBehaviourAction::NotifyHandler {
+                            peer_id,
+                            handler: NotifyHandler::One(connection),
+                            event: InEvent::DiscoverResponse {
+                                discovered: discovered.clone(),
+                                cookie,
+                            },
+                        },
+                        NetworkBehaviourAction::GenerateEvent(Event::DiscoverServed {
+                            enquirer: peer_id,
+                            registrations: discovered,
+                        }),
+                    ]
+                }
+                Err(_) => {
+                    let error = ErrorCode::InvalidCookie;
+
+                    vec![
+                        NetworkBehaviourAction::NotifyHandler {
+                            peer_id,
+                            handler: NotifyHandler::One(connection),
+                            event: InEvent::DeclineDiscoverRequest(error),
+                        },
+                        NetworkBehaviourAction::GenerateEvent(Event::DiscoverNotServed {
+                            enquirer: peer_id,
+                            error,
+                        }),
+                    ]
+                }
+            },
+            OutEvent::Discovered {
+                registrations,
+                cookie,
+            } => {
+                vec![NetworkBehaviourAction::GenerateEvent(Event::Discovered {
+                    rendezvous_node: peer_id,
+                    registrations,
+                    cookie,
+                })]
+            }
+            OutEvent::DiscoverFailed { namespace, error } => {
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::DiscoverFailed {
+                        rendezvous_node: peer_id,
+                        namespace,
+                        error,
+                    },
+                )]
+            }
+        };
+
+        self.events.extend(new_events);
```

Comment:
I had plans to extract this mapping into a separate function and unit test it but never got around to it :)

I find it somewhat cleaner to produce a list of new events as part of handling the incoming one instead of each branch causing side-effects. I agree though that it is somewhat error prone. Maybe extracting a function is good enough to enforce the return value of a `Vec` of events?

Reply:
> I had plans to extract this mapping into a separate function

Done as part of https://github.com/libp2p/rust-libp2p/pull/2107/commits/8f46361d90c95ea340bb95c5edd7aa60324e2a75.